### PR TITLE
fix(synthetic-shadow): synthetic shadow doesn't preserve assignedSlot behavior

### LINKED ISSUE #2377: Synthetic shadow doesn't preserve assignedSlot behavior
### Description

After loading the `@lwc/synthetic-shadow` polyfill, accessing [`assignedSlot`](https://developer.mozilla.org/en-US/docs/Web/API/Element/assignedSlot) for elements assigned to slots created outside of LWC always return `null`. The synthetic shadow polyfill should not interfere with native shadow DOM usage outside of LWC.

#### Steps to Reproduce

[Minimal repro](https://jsfiddle.net/kvo3pc2w/4/):

```js
const div = document.createElement('div');
document.body.appendChild(div);

div.attachShadow({ mode: 'open' }).innerHTML = `
	<slot></slot>
`;

const slotted = document.createElement('div')
slotted.textContent = 'slotted'
div.appendChild(slotted)

console.log('expected a slot, got:', slotted.assignedSlot)
```

#### Expected Results

`expected a slot, got: <slot>`

#### Actual Results

`expected a slot, got: null`

### Browsers Affected

All

### Version

- LWC: 2.2.2

### PR DESCRIPTION
**Pull request by rui-rayqiu**
## Details
Fixes #2377 

When synthetic shadow polyfill is loaded, accessing `assignedSlot` for elements that are assigned to slots inside native shadow components will return null. The proposed fix is to call the native `assignedSlot` getter under this condition.

## Does this pull request introduce a breaking change?

* ✅ No, it does not introduce a breaking change.

## Does this pull request introduce an observable change?

* ✅ No, it does not introduce an observable change.

## GUS work item

W-9492998

### CODE REVIEW

**Review by ekashida — CHANGES_REQUESTED** [review]
I'm not sure this is the right logic, I think we want something more like this:
```
const nodeOwnerKey = getNodeOwnerKey(this);
const isSyntheticShadowed = !isUndefined(nodeOwnerKey);
if (isSyntheticShadowed) {
    // existing logic for synthetic goes here
} else {
    if (this instanceof Element) {
        return originalElementAssignedSlotGetter.call(this); 
    }
    if (this instanceof Text) {
        return originalTextAssignedSlotGetter.call(this);
    }
}

return null;
```

Could you also make sure we have sufficient test coverage for this? At the least, we should have the following test cases:

- default slotted content should return `null` for `assignedSlot` for light, synthetic, and native (see what current behavior is for synthetic before making changes since it might be breaking)
- `assignedSlot` should return the expected slot for light, synthetic, and native

File: packages/@lwc/synthetic-shadow/src/faux-shadow/slot.ts [hunk 1]
```diff
@@ -83,9 +89,20 @@ function getFilteredSlotFlattenNodes(slot: HTMLElement): Node[] {
...
-export function assignedSlotGetterPatched(this: Element): HTMLSlotElement | null {
+export function assignedSlotGetterPatched(this: Element | Text): HTMLSlotElement | null {
     const parentNode = parentNodeGetter.call(this);
 
+    // use original assignedSlot if parent has a native shdow root
+    if (parentNode instanceof Element) {
+        const sr = parentNode.shadowRoot;
+        if (sr && isInstanceOfNativeShadowRoot(sr)) {
+            if (this instanceof Text) {
+                return originalTextAssignedSlotGetter.call(this);
```
[the remark this answers is not in the document]
Comment (rui-rayqiu) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/slot.ts:100]: @ekashida I see [here](https://github.com/salesforce/lwc/blob/d93e2963d8d32b138ce9ed130d98e7bef84b3d74/packages/%40lwc/synthetic-shadow/src/faux-shadow/text.ts#L15) that we put the same function for Text node so it should be reachable. We have a test [here](https://github.com/salesforce/lwc/blob/30ca15d645c88e232905d782ac1163f489a596a8/packages/integration-karma/test/shadow-dom/Element-properties/Element.assignedSlot.spec.js#L63) for slotted text node.

**Comment by rui-rayqiu** [comment]
> isSyntheticShadowed



> I'm not sure this is the right logic, I think we want something more like this:
> 
> ```
> const nodeOwnerKey = getNodeOwnerKey(this);
> const isSyntheticShadowed = !isUndefined(nodeOwnerKey);
> if (isSyntheticShadowed) {
>     // existing logic for synthetic goes here
> } else {
>     if (this instanceof Element) {
>         return originalElementAssignedSlotGetter.call(this); 
>     }
>     if (this instanceof Text) {
>         return originalTextAssignedSlotGetter.call(this);
>     }
> }
> 
> return null;
> ```
> 
> Could you also make sure we have sufficient test coverage for this? At the least, we should have the following test cases:
> 
> * default slotted content should return `null` for `assignedSlot` for light, synthetic, and native (see what current behavior is for synthetic before making changes since it might be breaking)
> * `assignedSlot` should return the expected slot for light, synthetic, and native

**Comment by rui-rayqiu** [comment]
> I'm not sure this is the right logic, I think we want something more like this:
> 
> ```
> const nodeOwnerKey = getNodeOwnerKey(this);
> const isSyntheticShadowed = !isUndefined(nodeOwnerKey);
> if (isSyntheticShadowed) {
>     // existing logic for synthetic goes here
> } else {
>     if (this instanceof Element) {
>         return originalElementAssignedSlotGetter.call(this); 
>     }
>     if (this instanceof Text) {
>         return originalTextAssignedSlotGetter.call(this);
>     }
> }
> 
> return null;
> ```
> 
> Could you also make sure we have sufficient test coverage for this? At the least, we should have the following test cases:
> 
> * default slotted content should return `null` for `assignedSlot` for light, synthetic, and native (see what current behavior is for synthetic before making changes since it might be breaking)
> * `assignedSlot` should return the expected slot for light, synthetic, and native

@ekashida I tried this logic but it does not work for light dom container. If the slotted content is a light DOM node, it doesn't have an owner key so `getNodeOwnerKey` will return undefined for this case too. Is there a suggested way to distinguish this case?

File: packages/@lwc/synthetic-shadow/src/faux-shadow/slot.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (ekashida) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/slot.ts:100]: I see, we are checking to see if the parent is an instance of `Element` and not the element itself.

File: packages/@lwc/synthetic-shadow/src/faux-shadow/slot.ts [hunk 2]
```diff
@@ -83,9 +89,20 @@ function getFilteredSlotFlattenNodes(slot: HTMLElement): Node[] {
...
 }
 
-export function assignedSlotGetterPatched(this: Element): HTMLSlotElement | null {
+export function assignedSlotGetterPatched(this: Element | Text): HTMLSlotElement | null {
     const parentNode = parentNodeGetter.call(this);
 
+    // use original assignedSlot if parent has a native shdow root
+    if (parentNode instanceof Element) {
+        const sr = parentNode.shadowRoot;
+        if (sr && isInstanceOfNativeShadowRoot(sr)) {
```
Comment (ekashida) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/slot.ts:98]: `isInstanceOfNativeShadowRoot` can handle `null` so we don't need the extra check.
```suggestion
        if (isInstanceOfNativeShadowRoot(sr)) {
```

File: packages/@lwc/synthetic-shadow/src/faux-shadow/slot.ts [hunk 3]
```diff
@@ -83,9 +89,20 @@ function getFilteredSlotFlattenNodes(slot: HTMLElement): Node[] {
     );
 }
 
-export function assignedSlotGetterPatched(this: Element): HTMLSlotElement | null {
+export function assignedSlotGetterPatched(this: Element | Text): HTMLSlotElement | null {
     const parentNode = parentNodeGetter.call(this);
 
+    // use original assignedSlot if parent has a native shdow root
+    if (parentNode instanceof Element) {
+        const sr = parentNode.shadowRoot;
```
[the remark this answers is not in the document]
Comment (rui-rayqiu) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/slot.ts:97]: @ekashida How do we use `shadowRootGetter` here? It seems that `shadowRootGetterPatched` is already patched on `Element.shadowRoot` [here](https://github.com/salesforce/lwc/blob/0de7ef2685fd26f0ba7d56987c8562a880922fe8/packages/%40lwc/synthetic-shadow/src/faux-shadow/element.ts#L169) so using dot-notation will be calling `shadowRootGetterPatched`. Maybe I'm missing something?

[the remark this answers is not in the document]
Comment (ekashida) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/slot.ts:97]:
```
import { shadowRootGetter } from '../env/element';

...

if (parentNode instanceof Element) {
    const sr = shadowRootGetter.call(parentNode);
    if (isInstanceOfNativeShadowRoot(sr)) {
        ...
    }
}

[the remark this answers is not in the document]
Comment (rui-rayqiu) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/slot.ts]: I see. That makes sense. Revised accordingly. Thanks!

[the remark this answers is not in the document]
Comment (rui-rayqiu) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/slot.ts]: I think we have basic test coverage for [assignedSlot](https://github.com/salesforce/lwc/blob/master/packages/integration-karma/test/shadow-dom/Element-properties/Element.assignedSlot.spec.js) and for [light dom](https://github.com/salesforce/lwc/blob/ba58615566edc465b96b7c71eef0f6fb47e15b0f/packages/integration-karma/test/light-dom/synthetic-shadow/index.spec.js#L21). I've also did some manual tests on default slot content returning `null` for `assignedSlot`.

[the remark this answers is not in the document]
Comment (ekashida) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/slot.ts]: > I've also did some manual tests on default slot content returning null for assignedSlot.

Would be great to add the missing coverage for this.

**Comment by ekashida** [comment]
The failing FF test should be fixed after a rebase.

**Review by ekashida — APPROVED** [review]
